src/browserlib/extract-algorithms.mjs

```diff
@@ -0,0 +1,663 @@
+ * nested ones. Nested ones typically don't have names.
+ */
+function serializeAlgorithm(algo, context) {
+  let res = getAlgorithmInfo(algo, context);
+  res.rationale = algo.rationale;
+  res.steps = serializeSteps(algo.root);
+  return res;
+}
+
+/**
+ * Serialize the given steps contained in the given root element.
+ */
+function serializeSteps(root) {
+  if (root.nodeName === 'DL') {
+    return [
+      {
+        operation: 'switch',
+        steps: [...root.querySelectorAll('& > dt')].map(option => {
+          let dd = option.nextElementSibling;
+          while (dd && dd.nodeName !== 'DD') {
+            dd = dd.nextElementSibling;
+          }
+          if (!dd) {
+            throw new Error('Switch option without <dd> found: ' + option.textContent);
+          }
+          return {
+            'case': getTextContent(option),
+            steps: serializeSteps(dd)
+          };
+        })
+      }
+    ]
+  }
+  else if (root.nodeName === 'OL') {
+    return [...root.querySelectorAll('& > li')].map(serializeStep);
+  }
+  else {
+    return [serializeStep(root)];
+  }
+}
+
+/**
+ * Serialize an algorithm step
+ */
+function serializeStep(li) {
+  let res = {};
+  const candidateAlgorithms = findAlgorithms(li, { includeIgnored: true });
+  const algorithms = candidateAlgorithms.filter(algo => !!algo.rationale);
+  if (algorithms.length > 0) {
+    res = serializeAlgorithm(algorithms[0], { nested: true });
+  }
+  if (!res.html) {
+    res.html = getHTMLContent(li);
+  }
+  if (algorithms.length > 1) {
+    res.additional = algorithms.slice(1)
+      .map(algo => serializeAlgorithm(algo, { nested: true }));
+  }
+  const ignoredAlgorithms = candidateAlgorithms
+    .filter(algo => !algo.rationale)
+    .map(algo => getTextContent(algo.root));
+  if (ignoredAlgorithms.length > 0) {
+    res.ignored = ignoredAlgorithms;
+  }
+  return res;
+}
+
+/**
+ * Parse a list element looking for algorithmic operations or other anchors
+ * that should allow us to assess that the steps are indeed part of an
+ * algorithm. Return a string representation of that rationale.
+ */
+function findRationale(ol) {
+  let rationale = null;
+
+  if (ol.matches('.algorithm')) {
+    return '.algorithm';
+  }
+  const text = getTextContent(ol).toLowerCase();
+  rationale = stepOperations.find(op => {
+    return text.match(new RegExp(`^${op}(\\.|:| )`, 'i'));
+  });
+
+  if (!rationale) {
+    rationale = stepInlineOperations.find(op => {
+      if (typeof op === 'string') {
+        return text.includes(op);
+      }
+      else {
+        return text.match(op);
+      }
+    });
+  }
+
+  if (!rationale) {
+    rationale = stepAnchors.find(anchor => {
+      if (typeof anchor === 'string') {
+        return text.includes(anchor);
+      }
+      else {
+        return text.match(anchor);
+      }
+    });
+  }
+
+  return rationale?.toString();
+}
+
+
+/**
+ * Return true if element has the given parent as ancestor
+ */
+function hasParent(el, parent) {
```

Comment:
this sounds like a rewrite of [`Node.contains()`](https://developer.mozilla.org/en-US/docs/Web/API/Node/contains)

Reply:
Ah yes, I had forgotten that function existed ;)